### PR TITLE
[PP] Let PP split BlockMask into micro-BlockMask

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #163185
* #164500
* #162542
* __->__ #164111

BlockMask has batch dimension information. So PP has to split it as well just like all other tensors. All the tensors in BlockMask have the batch dimension, so we can just split it without too many issues. However, `mask_mod` requires the batch index as the input, which the value is going to be changed after the split. So we have to wrap it inside a closure to modify the batch index.

cc @H-Huang @awgu @wanchaol @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

### CODE REVIEW

File: test/distributed/pipelining/test_microbatch.py [hunk 1]
```diff
@@ -59,6 +61,82 @@ def test_split_and_merge(self):
...
+                return (seq_idx[b, q_idx] == seq_idx[b, kv_idx]) & (q_idx >= kv_idx)
+
+            return block_causal_mask
+
+        batch = list(range(30)) * 5
+        batch = torch.tensor(batch[: B * Q_LEN], device=device).reshape(B, Q_LEN)
+        q = torch.randn(B, H, Q_LEN, DIM, device=device)
+        k = torch.randn(B, H, KV_LEN, DIM, device=device)
+        v = torch.randn(B, H, KV_LEN, DIM, device=device)
+        # block_mask_fn = torch.compile(create_block_mask)
```
Comment (XilunWu) [inline test/distributed/pipelining/test_microbatch.py]: is compiled create_block_mask causing issue?

File: test/distributed/pipelining/test_microbatch.py [hunk 2]
```diff
@@ -59,6 +61,82 @@ def test_split_and_merge(self):
...
+        block_mask_fn = create_block_mask
+        block_mask = block_mask_fn(
+            create_block_causal_mask(batch, 29),
+            B=B,
+            H=H,
+            Q_LEN=Q_LEN,
+            KV_LEN=KV_LEN,
+            device=device,
+        )
+        # flex_fn = torch.compile(flex_attention)
```
Comment (XilunWu) [inline test/distributed/pipelining/test_microbatch.py]: ditto

File: torch/distributed/pipelining/microbatch.py [hunk 3]
```diff
@@ -132,112 +209,58 @@ def _shard_dict_of_args(
...
+        else:
             raise ValueError(
-                f"Argument value {arg} did not have the same number of "
-                f"values as as chunk spec {chunk_spec}"
+                f"Unsupported chunk spec: {spec} and value: {v} combination."
+            )
+    num_chunks = min(min(split_sizes), num_chunks)
+
+    flat_split_results = [[] for _ in range(num_chunks)]
+    for v, spec in zip(values, chunk_specs):
```
Comment (Skylion007) [inline torch/distributed/pipelining/microbatch.py]: We are on Python 3.10 now officially, use strict=True kwarg for zip

File: test/distributed/pipelining/test_microbatch.py [hunk 4]
```diff
@@ -59,6 +61,82 @@ def test_split_and_merge(self):
         torch.testing.assert_close(merged_kwargs, kwargs)
         print("Microbatch test passed")
 
+    def test_split_block_mask(self, device):
+        B = 4
+        H = 1
+        Q_LEN = 32
+        KV_LEN = 32
```
Comment (XilunWu) [inline test/distributed/pipelining/test_microbatch.py]: this size could be too small for testing, in the result `block_mask` there's only one block of size [128, 128].

File: test/distributed/pipelining/test_microbatch.py [hunk 5]
```diff
@@ -59,6 +61,82 @@ def test_split_and_merge(self):
...
+            args_chunk_spec=None,
+            kwargs_chunk_spec=None,
+        )
+
+        for i in range(B):
+            q_chunk, k_chunk, v_chunk, block_mask_chunk = arg_split[i]
+            out_chunk = flex_fn(q_chunk, k_chunk, v_chunk, block_mask=block_mask_chunk)
+            self.assertEqual(q_chunk.squeeze(0), q[i])
+            self.assertEqual(k_chunk.squeeze(0), k[i])
+            self.assertEqual(v_chunk.squeeze(0), v[i])
```
Comment (XilunWu) [inline test/distributed/pipelining/test_microbatch.py]: would only hold true if `batch_size == world_size`.

**Review by XilunWu — COMMENTED** [review]
reviewed the test. Reviewing the implementation now.

File: torch/distributed/pipelining/microbatch.py (hunk 3 above)
Comment (XilunWu) [reply torch/distributed/pipelining/microbatch.py]: thanks for the reminder ;-)

File: torch/distributed/pipelining/microbatch.py [hunk 6]
```diff
@@ -115,6 +116,82 @@ class _Replicate:
...
+        ret.append(
+            BlockMask.from_kv_blocks(
+                kv_num_blocks=kv_num_blocks_chunks[chunk_idx],
+                kv_indices=kv_indices_chunks[chunk_idx],
+                full_kv_num_blocks=full_kv_num_blocks_chunks[chunk_idx],
+                full_kv_indices=full_kv_indices_chunks[chunk_idx],
+                BLOCK_SIZE=block_mask.BLOCK_SIZE,
+                mask_mod=create_mask_mod(batch_offset),
+                seq_lengths=block_mask.seq_lengths,
+            )
```
Comment (XilunWu) [inline torch/distributed/pipelining/microbatch.py:173]: hey @drisspg , do you happen to know how performant this API call is compared to calling (compiled) `create_block_mask`? 

Chien-chin constructs the new `block_mask` in a different way than the one in `create_cp_block_mask` because the rewriting logic is much simpler than CP, but I still hope to unify these 2 functions (`split_block_mask` and `create_cp_block_mask`) since they're doing similar things.

File: torch/distributed/pipelining/microbatch.py [hunk 7]
```diff
@@ -132,112 +209,58 @@ def _shard_dict_of_args(
...
+    ), f"args_dict.keys() = {list(args_dict.keys())} args_chunk_spec.keys() = {list(args_chunk_spec.keys())}"
+    assert args_chunk_spec is not None  # Should have been set by caller
+
+    values, tree_spec = tree_flatten(args_dict)
+    chunk_specs, _ = tree_flatten(args_chunk_spec)
+    assert len(values) == len(chunk_specs)
+
+    # Fist check and find the actual number of chunks
+    split_sizes = []
+    for v, spec in zip(values, chunk_specs):
```
Comment (XilunWu) [inline torch/distributed/pipelining/microbatch.py]: as Aaron pointed out we should add `strict=True`

**Review by XilunWu — COMMENTED** [review]
Overall LGTM. Let me know when you need a final review.

File: torch/distributed/pipelining/microbatch.py [hunk 8]
```diff
@@ -132,112 +209,58 @@ def _shard_dict_of_args(
...
+            assert isinstance(spec, TensorChunkSpec)
+            assert spec.split_dim == 0
+            split_sizes.append(v.kv_num_blocks.size(0))
+        else:
             raise ValueError(
-                f"Argument value {arg} did not have the same number of "
-                f"values as as chunk spec {chunk_spec}"
+                f"Unsupported chunk spec: {spec} and value: {v} combination."
+            )
+    num_chunks = min(min(split_sizes), num_chunks)
```
Comment (H-Huang) [inline torch/distributed/pipelining/microbatch.py]: nit: call it `result_num_chunks`?

File: torch/distributed/pipelining/microbatch.py [hunk 9]
```diff
@@ -132,112 +209,58 @@ def _shard_dict_of_args(
...
+    split_sizes = []
+    for v, spec in zip(values, chunk_specs):
+        if spec is _Replicate:
+            split_sizes.append(num_chunks)
+        elif isinstance(v, torch.Tensor):
+            assert isinstance(spec, TensorChunkSpec)
+            split_sizes.append(v.size(spec.split_dim))
+        elif isinstance(v, BlockMask):
+            assert isinstance(spec, TensorChunkSpec)
+            assert spec.split_dim == 0
```
Comment (H-Huang) [inline torch/distributed/pipelining/microbatch.py]: add assert error message

**Review by H-Huang — COMMENTED** [review]
Overall changes look reasonable to me, will do another pass soon to understand the tests better

File: torch/distributed/pipelining/microbatch.py (hunk 6 above)
Comment (fegin) [reply torch/distributed/pipelining/microbatch.py:173]: From @drisspg `from_kv_blocks` is strictly faster.

File: torch/distributed/pipelining/microbatch.py [hunk 10]
```diff
@@ -132,114 +238,58 @@ def _shard_dict_of_args(
...
-                        new_val[slice_indices] = chunk_tensor
-
-                        expanded_chunks.append(new_val)
-
-                        split_dim_idx += chunk_tensor.size(chunk_v.split_dim)
-
-                    sharded_arg_flat.append(expanded_chunks)
-                else:
-                    sharded_arg_flat.append(chunk_tensors)  # type: ignore[arg-type]
+        for _flat_split_result, _v_split in zip(flat_split_results, v_splits):
```
Comment (Skylion007) [inline torch/distributed/pipelining/microbatch.py]: Should this be strict=True or strict=False?

File: test/distributed/pipelining/test_microbatch.py [hunk 11]
```diff
@@ -59,6 +60,120 @@ def test_split_and_merge(self):
...
+                h: torch.Tensor,
+                q_idx: torch.Tensor,
+                kv_idx: torch.Tensor,
+            ):
+                return (seq_idx[b, q_idx] == seq_idx[b, kv_idx]) & (q_idx >= kv_idx)
+
+            return block_causal_mask
+
+        batch = list(range(DOC_LEN)) * 1024
+        batch = torch.tensor(batch[: B * SEQ_LEN], device=device).reshape(B, SEQ_LEN)
```
Comment (drisspg) [inline test/distributed/pipelining/test_microbatch.py]: nit: this is a little weird to me, I think at least do comment or  have some assert that doc_len * 1024 > b * seqlen

File: test/distributed/pipelining/test_microbatch.py [hunk 12]
```diff
@@ -59,6 +60,120 @@ def test_split_and_merge(self):
...
+            B=B,
+            H=H,
+            Q_LEN=SEQ_LEN,
+            KV_LEN=SEQ_LEN,
+            device=device,
+        )
+        if device == "cuda":
+            flex_fn = torch.compile(flex_attention)
+        else:
+            # It's unclear why CPU + torch.compile + flex_attention can cause an issue.
```
Comment (drisspg) [inline test/distributed/pipelining/test_microbatch.py:108]: cpu flex code path in inductor is not as battle tested
and has poor training support AFAIK

File: test/distributed/pipelining/test_microbatch.py [hunk 13]
```diff
@@ -59,6 +60,120 @@ def test_split_and_merge(self):
...
+        k_chunks = []
+        dk_chunks = []
+        v_chunks = []
+        dv_chunks = []
+        block_mask_chunks = []
+        out_chunks = []
+        for i in range(len(arg_split)):
+            q_chunk, k_chunk, v_chunk, block_mask_chunk = arg_split[i]
+            for chunk, chunks in zip(
+                (q_chunk, k_chunk, v_chunk), (q_chunks, k_chunks, v_chunks)
```
Comment (drisspg) [inline test/distributed/pipelining/test_microbatch.py]: nit this is a lil hard to understand, maybe q_chunk and q_total

File: torch/distributed/pipelining/microbatch.py [hunk 14]
```diff
@@ -115,6 +116,111 @@ class _Replicate:
     pass
 
 
+def _split_block_mask(
+    block_mask: BlockMask,
+    spec: TensorChunkSpec,
```
Comment (drisspg) [inline torch/distributed/pipelining/microbatch.py]: Is TensorChunkSpec a distributed API already or new?

File: torch/distributed/pipelining/microbatch.py [hunk 15]
```diff
@@ -115,6 +116,111 @@ class _Replicate:
...
+    num_chunks: int,
+) -> list[BlockMask]:
+    """Given a block mask, and a chunking spec, split the block mask.
+
+    Args:
+        block_mask: Block mask to split
+        spec: Chunking spec
+        num_chunks: Number of chunks to split the block mask into
```
Comment (drisspg) [inline torch/distributed/pipelining/microbatch.py:127]: maybe add a note somehwhere only batch dims support splitting?

File: torch/distributed/pipelining/microbatch.py [hunk 16]
```diff
@@ -115,6 +116,111 @@ class _Replicate:
...
+    Returns:
+        chunk_block_masks: List of chunked block masks
+    """
+
+    assert block_mask.kv_num_blocks.size(0) >= num_chunks, (
+        "Block mask has fewer batch size than the number of chunks. "
+    )
+
+    kv_num_blocks_chunks = torch.tensor_split(
+        block_mask.kv_num_blocks, num_chunks, spec.split_dim
```
Comment (drisspg) [inline torch/distributed/pipelining/microbatch.py]: needs to be dim 0 right?

File: torch/distributed/pipelining/microbatch.py (hunk 15 above)
Comment (drisspg) [inline torch/distributed/pipelining/microbatch.py:127]: Is num_chunks not embedded in the chunk spec?

File: torch/distributed/pipelining/microbatch.py [hunk 17]
```diff
@@ -115,6 +116,111 @@ class _Replicate:
...
+    for chunk_idx in range(num_chunks):
+
+        def create_mask_mod(idx):
+            def mask_mod_wrapper(b, h, q_idx, kv_idx):
+                b_offset = torch.full_like(b, idx)
+                return block_mask.mask_mod(b + b_offset, h, q_idx, kv_idx)
+
+            return mask_mod_wrapper
+
+        # `from_kv_blocks` is strictly faster than `create_block_mask`.
```
Comment (drisspg) [inline torch/distributed/pipelining/microbatch.py]: nit, I dont think you need to say this,  if you have the final tensors you dont need to go through and apply vmap + mask-mod to build them from scratch. So its more like create_block_mask is a factory and it ends up calling from_kv_blocks which is the the base constructor

File: torch/distributed/pipelining/microbatch.py [hunk 18]
```diff
@@ -115,6 +116,111 @@ class _Replicate:
...
+        if block_mask.full_kv_indices is not None
+        else [None] * num_chunks
+    )
+
+    chunk_block_masks = []
+    batch_offset = 0
+    for chunk_idx in range(num_chunks):
+
+        def create_mask_mod(idx):
+            def mask_mod_wrapper(b, h, q_idx, kv_idx):
```
Comment (drisspg) [inline torch/distributed/pipelining/microbatch.py]: nit: maybe call this a batch offset mask mod

File: test/distributed/pipelining/test_microbatch.py (hunk 11 above)
Comment (H-Huang) [inline test/distributed/pipelining/test_microbatch.py]: can do this instead:

```
total_elements = B * SEQ_LEN
batch = torch.arange(total_elements, device=device) % DOC_LEN
batch = batch.reshape(B, SEQ_LEN)
```

File: torch/distributed/pipelining/microbatch.py [hunk 19]
```diff
@@ -132,114 +238,58 @@ def _shard_dict_of_args(
...
-        chunk_spec = args_chunk_spec[arg_key]
-        assert chunk_spec is not None  # Should have been set by caller
-        chunk_spec_flat, _ = tree_flatten(chunk_spec)
-        if len(flat) != len(chunk_spec_flat):
+    assert args_chunk_spec is not None  # Should have been set by caller
+
+    values, tree_spec = tree_flatten(args_dict)
+    chunk_specs, _ = tree_flatten(args_chunk_spec)
+
+    # Fist check and find the actual number of chunks
```
Comment (H-Huang) [inline torch/distributed/pipelining/microbatch.py]:
```suggestion
    # First check and find the actual number of chunks
```

File: torch/distributed/pipelining/microbatch.py (hunk 14 above)
Comment (H-Huang) [reply torch/distributed/pipelining/microbatch.py]: distributed API, it specifies which dimension of a tensor to split along when dividing it into chunks for microbatching (in pipeline parallelism).

File: torch/distributed/pipelining/microbatch.py [hunk 20]
```diff
@@ -115,6 +116,111 @@ class _Replicate:
...
+def _split_block_mask(
+    block_mask: BlockMask,
+    spec: TensorChunkSpec,
+    num_chunks: int,
+) -> list[BlockMask]:
+    """Given a block mask, and a chunking spec, split the block mask.
+
+    Args:
+        block_mask: Block mask to split
+        spec: Chunking spec
```
Comment (H-Huang) [inline torch/distributed/pipelining/microbatch.py]: is there a case where we would want to specify on non-0 dimension for block mask? If it's always 0, then maybe passing the spec is not necessary

**Review by H-Huang — APPROVED** [review]
Looks good! Thanks for cleaning up `_shard_dict_of_args()`, it is much more understandable now.

File: torch/distributed/pipelining/microbatch.py (hunk 15 above)
Comment (fegin) [reply torch/distributed/pipelining/microbatch.py:127]: No, chunk spec only specifies the chunk dim.

File: torch/distributed/pipelining/microbatch.py (hunk 20 above)
Comment (fegin) [reply torch/distributed/pipelining/microbatch.py]: sound good

File: torch/distributed/pipelining/microbatch.py (hunk 16 above)
Comment (fegin) [reply torch/distributed/pipelining/microbatch.py]: Right, I checked it beforehand. But maybe just like @H-Huang  mentioned, just remove chunk spec for this function at all.

File: torch/distributed/pipelining/microbatch.py (hunk 10 above)
Comment (fegin) [reply torch/distributed/pipelining/microbatch.py]: True.

File: torch/distributed/pipelining/microbatch.py [hunk 21]
```diff
@@ -132,114 +234,60 @@ def _shard_dict_of_args(
...
+        v_splits: Sequence[Any] = []
+        if spec is _Replicate:
+            v_splits = [v] * result_num_chunks
+        elif isinstance(v, torch.Tensor):
+            v_splits = _split_tensor(v, spec, result_num_chunks)
+        elif isinstance(v, BlockMask):
+            v_splits = _split_block_mask(v, result_num_chunks)
+        else:
+            raise ValueError(
+                f"Unsupported chunk spec: {spec} and value: {v} combination."
```
Comment (carmocca) [inline torch/distributed/pipelining/microbatch.py:279]: This raises for primitive type returns. For example if the model forward returns a boolean for whether to materialize logits

Comment (fegin) [reply torch/distributed/pipelining/microbatch.py:279]: Sorry, I don't understand the comment. Does this raise cause any issue? @carmocca

Comment (carmocca) [reply torch/distributed/pipelining/microbatch.py:279]: Well it now fails for primitive (e.g. boolean) return values, when they used to work.

I worked around it by explicitly marking it as a `_Replicate`
```py
        pp_schedule._kwargs_chunk_spec = {
            "a_tensor_return": TensorChunkSpec(0),
            "a_tensor_return": TensorChunkSpec(0),
            "a_boolean_return": _Replicate,  # type: ignore[dict-item]
        }
```

Comment (fegin) [reply torch/distributed/pipelining/microbatch.py:279]: @carmocca I understand now. I believe my follow-up PR fixes it, https://github.com/pytorch/pytorch/pull/165306. Let me know if you still encounter the error with https://github.com/pytorch/pytorch/pull/165306.
